### PR TITLE
Add and, implies, and partial forall support to assertion parser

### PR DESCRIPTION
**Pull request by karlosss**

### CODE REVIEW

File: new/prusti-specs/src/specifications/common.rs [hunk 1]
```diff
@@ -256,6 +256,12 @@ pub struct ProcedureSpecification<EID, ET, AT> {
     pub posts: Vec<Assertion<EID, ET, AT>>,
 }
 
+#[derive(Debug, Clone)]
+pub struct Arg {
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/common.rs]: This cannot be shared between typed and untyped specifications and, therefore, belongs to untyped ones.

File: new/prusti-specs/src/specifications/preparser.rs [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+        }
+    }
+    fn from_token_stream(tokens: TokenStream) -> Self {
+        let token_queue: VecDeque<_> = tokens.into_iter().collect();
+        Self {
+            tokens: token_queue,
+            last_span: Span::call_site(),
+        }
+    }
+
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]: Nit: Please make the spacing between functions consistent.

File: new/prusti-specs/src/specifications/preparser.rs [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+    /// Check if there is a subexpression that contains both and and or
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]:
```suggestion
    /// Check if there is a subexpression that contains both `&&` and `||`.
```
Could you also explain in the comment why is this needed?

File: new/prusti-specs/src/specifications/preparser.rs [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+    fn contains_both_and_or(&mut self) -> bool {
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs:47]: As far as I understand, this method is not supposed to modify `self`, right? If yes, it should take `&self`, not `&mut self`. It would be best to iterate over the tokens without copying them, if possible. Otherwise, you could create a new instance of `ParserStream`.

File: new/prusti-specs/src/specifications/preparser.rs [hunk 5]
```diff
@@ -0,0 +1,584 @@
+use proc_macro2::{Delimiter, Group, Spacing, Span, TokenStream, TokenTree, Ident, };
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]: Could you please add a high-level explanation of what is the purpose of this file?

File: new/prusti-specs/src/specifications/preparser.rs [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+                if let Some(TokenTree::Group(group)) = self.pop() {
+                    return Some(group);
+                } else {
+                    unreachable!();
+                }
+            }
+        }
+        None
+    }
+
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]: Please make the spacing between methods consistent.

File: new/prusti-specs/src/specifications/preparser.rs [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+        while !self.is_empty() {
+            t.push(self.pop().unwrap());
+        }
+        stream.extend(t.into_iter());
+        stream
+    }
+}
+
+#[derive(Debug)]
+struct OneArg {
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]: This is `OneArg` for what? Is it for `forall`? If yes, maybe `ForallArg` would be a more suitable name?

File: new/prusti-specs/src/specifications/preparser.rs [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+        Ok(Self{
+            name: input.parse()?,
+            colon: input.parse()?,
+            typ: input.parse()?
+        })
+    }
+}
+
+#[derive(Debug)]
+struct AllArgs {
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]: Same comment as for `OneArg` regarding the name.

File: new/prusti-specs/src/specifications/preparser.rs [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+///
+/// The following are also Prusti assertions:
+/// `forall(|NAME1: TYPE1, NAME2: TYPE2, ...| A)`
+/// `forall(|NAME1: TYPE1, NAME2: TYPE2, ...| A, triggers=[(E, ...), ...])`
+///
+/// Prusti assertions can only be joined together by `&&` and `==>`, for example the following
+/// is not allowed, since `(E ==> E)` is a Prusti assertion:
+/// `(E ==> E) || E`
+/// However, this is fine, as `&&` is also a Rust operator:
+/// `(E && E) && E`
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]: Didn't you intend to have `(E && E) || E`?

File: new/prusti-specs/src/specifications/preparser.rs [hunk 10]
```diff
@@ -0,0 +1,584 @@
...
+            input: input,
+            conjuncts: Vec::new(),
+            expr: Vec::new(),
+            previous_expression_resolved: false,
+            expected_operator: false,
+            expected_only_operator: false
+        }
+    }
+
+    fn from_parser_stream(input: &mut ParserStream) -> Self {
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]: I think it would be cleaner to take by value `input: ParserStream` and do `mem::replace` on the caller side.

File: new/prusti-specs/src/specifications/preparser.rs [hunk 11]
```diff
@@ -0,0 +1,584 @@
...
+            conjuncts: Vec::new(),
+            expr: Vec::new(),
+            previous_expression_resolved: false,
+            expected_operator: false,
+            expected_only_operator: false
+        }
+    }
+
+    /// Creates a single Prusti assertion from the input and returns it.
+    pub fn extract_assertion(&mut self) -> syn::Result<AssertionWithoutId> {
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]: This method is huge. Could you split it in several smaller ones?

**Review by vakaras — COMMENTED** [review]
I did the first pass.

Two main requests:
1. Could you add a high-level explanation of how the parsing is supposed to work at the beginning of `new/prusti-specs/src/specifications/mod.rs` ?
2. Could you add tests that check the new functionality?

File: new/prusti-specs/src/specifications/common.rs (hunk 1 above)
Comment (karlosss) [reply new/prusti-specs/src/specifications/common.rs]: I then need something like `untyped_common.rs`, recursive imports otherwise

File: new/prusti-specs/src/specifications/preparser.rs (hunk 9 above)
Comment (karlosss) [reply new/prusti-specs/src/specifications/preparser.rs]: yes I did, thanks!

File: new/prusti-specs/src/specifications/preparser.rs (hunk 4 above)
Comment (karlosss) [reply new/prusti-specs/src/specifications/preparser.rs:47]: this needs to set the `last_span` as well. So it has to be mutable.

File: new/prusti-specs/src/specifications/mod.rs [hunk 12]
```diff
@@ -1,5 +1,11 @@
+/// The parsing of Prusti assertions consists of the following steps:
+/// - Split the input into atomic Prusti assertions (using operators `&&` and `==>` and `forall` keyword).
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/mod.rs]: Could you please define *atomic Prusti assertion*?

File: new/prusti-specs/src/specifications/preparser.rs [hunk 13]
```diff
@@ -0,0 +1,605 @@
+/// The preparser splits a composite Prusti assertion into atomic assertions, parses the resulting
+/// Rust expressions, and then assembles the composite Prusti assertion.
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]: Could you explain the relation between *atomic assertion* and *Rust expression*.

File: new/prusti-specs/src/specifications/preparser.rs [hunk 14]
```diff
@@ -0,0 +1,605 @@
...
+        while !self.is_empty() {
+            t.push(self.pop().unwrap());
+        }
+        stream.extend(t.into_iter());
+        stream
+    }
+}
+
+#[derive(Debug)]
+struct ForAllArg {
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]: Why do you need both `Arg` and `ForAllArg`, which are basically identical?

File: new/prusti-specs/src/specifications/untyped.rs [hunk 15]
```diff
@@ -86,6 +83,39 @@ impl AssignExpressionId<Expression> for common::Expression<(), syn::Expr> {
...
+}
+
+impl AssignExpressionId<TriggerSet> for common::TriggerSet<(), syn::Expr> {
+    fn assign_id(
+        self,
+        spec_id: SpecificationId,
+        id_generator: &mut ExpressionIdGenerator,
+    ) -> TriggerSet {
+        TriggerSet {
+            0: self.0
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/untyped.rs:106]:
```suggestion
  TriggerSet(self.0
```

File: new/prusti/tests/pass/parse/composite_fail.stderr [hunk 16]
```diff
@@ -0,0 +1,14 @@
+error: found || and && in the same subexpression
+  --> $DIR/composite_fail.rs:10:17
+   |
+10 | #[requires(true && true || true)]
+   |                 ^^^^^^^^^^
+
+error: expected Prusti operator
```
Comment (vakaras) [inline new/prusti/tests/pass/parse/composite_fail.stderr]:
```suggestion
error: expected Prusti operator (`&&` or `==>`)
```

File: new/prusti/tests/pass/parse/forall_fail.stderr [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+22 | #[requires(forall(|a, b| true))]
+   |                     ^
+
+error: expected Prusti operator
+  --> $DIR/forall_fail.rs:25:23
+   |
+25 | #[requires(forall(||) || forall(||))]
+   |                       ^^
+
+error: expected tuple
```
Comment (vakaras) [inline new/prusti/tests/pass/parse/forall_fail.stderr]: This error message is confusing.

File: new/prusti/tests/pass/parse/forall_fail.stderr [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+  --> $DIR/forall_fail.rs:28:18
+   |
+28 | #[requires(forall(|| 1+1 == 1+1, triggers=[1]))]
+   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: expected tuple
+  --> $DIR/forall_fail.rs:31:18
```
Comment (vakaras) [inline new/prusti/tests/pass/parse/forall_fail.stderr]: Same here.

File: new/prusti/tests/pass/parse/implies_fail.stderr [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+10 | #[requires(true ==>)]
+   |                 ^^^
+
+error: expected Prusti assertion
+  --> $DIR/implies_fail.rs:13:12
+   |
+13 | #[requires(==> true)]
+   |            ^^^
+
+error: expected expression
```
Comment (vakaras) [inline new/prusti/tests/pass/parse/implies_fail.stderr]:
```suggestion
error: expected expression (an implication cannot be inside a Rust expression)
```

Comment (karlosss) [reply new/prusti/tests/pass/parse/implies_fail.stderr]: this is not my error, this is the default error from `syn::parse`; is there a way to change it?

File: new/prusti/tests/pass/parse/forall_fail.stderr [hunk 20]
```diff
@@ -28,23 +28,29 @@ error: expected `:`
...
+   |                                           ^^^
 
-error: expected tuple
-  --> $DIR/forall_fail.rs:31:18
+error: `triggers` must be an array of tuples containing Rust expressions
+  --> $DIR/forall_fail.rs:31:37
    |
 31 | #[requires(forall(|| true, triggers=[(1,2), 1]))]
-   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |                                     ^^^^^^^^^^
```
Comment (vakaras) [inline new/prusti/tests/pass/parse/forall_fail.stderr]: Would it be possible to make the span more precise to point only to `1`?

File: new/prusti-specs/src/specifications/mod.rs [hunk 21]
```diff
@@ -1,5 +1,41 @@
...
+/// - `A && A` (conjunction)
+/// - `A ==> A` (implication)
+///
+/// `==>` has weaker binding than `&&` and is right-associative.
+///
+/// Parentheses can be used as usual, i.e. `(A ==> A) && A` is a Prusti assertion.
+///
+/// `forall` expression has the following syntax:
+/// `forall(|NAME1: TYPE1, NAME2: TYPE2, ...| A)`
+/// `forall(|NAME1: TYPE1, NAME2: TYPE2, ...| A, triggers=[(E, ...), ...])`
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/mod.rs:15]: I think it would be good to mention that the `forall` body is an assertion.

File: new/prusti-specs/src/specifications/mod.rs [hunk 22]
```diff
@@ -1,5 +1,41 @@
...
+/// are all illegal:
+/// `E && E || E`
+/// `A && (E && E || E)`
+///
+/// However, this is fine, since the implication resolves the ambiguity:
+/// (note that both the lhs and rhs of the implication form a Prusti assertion)
+/// `E && E ==> E || E`
+///
+/// Basic parser usage (`tokens` is of type `proc_macro2::TokenStream`):
+/// `let mut parser = Parser::from_token_stream(tokens);`
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/mod.rs]: Use this to have better rendering:

     ```rust
     code
     ```

File: new/prusti-specs/src/specifications/preparser.rs [hunk 23]
```diff
@@ -0,0 +1,637 @@
+/// The preparser splits a composite Prusti assertion into atomic assertions, parses the resulting
+/// Rust expressions, and then assembles the composite Prusti assertion.
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs]:
```suggestion
/// Rust expressions, and then assembles the composite Prusti assertion. For a more high-level overview, please check `mod.rs`.
```

(Do not forget to format the documentaiton lines to be no longer than 80 characters.)

File: new/prusti-specs/src/specifications/preparser.rs [hunk 24]
```diff
@@ -0,0 +1,637 @@
...
+use super::common;
+use crate::specifications::common::AssertionKind::ForAll;
+use crate::specifications::common::{ForAllVars, TriggerSet, Trigger};
+use syn::spanned::Spanned;
+
+pub type AssertionWithoutId = common::Assertion<(), syn::Expr, Arg>;
+pub type ExpressionWithoutId = common::Expression<(), syn::Expr>;
+
+#[derive(Debug, Clone)]
+struct ParserStream {
```
Comment (vakaras) [inline new/prusti-specs/src/specifications/preparser.rs:22]: Could you add a high-level explanation what is the purpose of this struct?

**Review by vakaras — CHANGES_REQUESTED** [review]

File: new/prusti-specs/src/specifications/preparser.rs (hunk 14 above)
Comment (karlosss) [reply new/prusti-specs/src/specifications/preparser.rs]: `ForAllArg` has that `:` token in it, which is a trash for further processing.

File: new/prusti-specs/src/specifications/untyped.rs (hunk 15 above)
Comment (karlosss) [reply new/prusti-specs/src/specifications/untyped.rs:106]: i will resolve this once this whole part of the code is refactored, as it is probably incorrect anyway

File: new/prusti-specs/src/specifications/preparser.rs (hunk 14 above)
Comment (vakaras) [reply new/prusti-specs/src/specifications/preparser.rs]: So, just drop that `:` in the parsing function.

File: new/prusti-specs/src/specifications/untyped.rs (hunk 15 above)
Comment (vakaras) [reply new/prusti-specs/src/specifications/untyped.rs:106]: Fine with me.

**Review by vakaras — APPROVED** [review]
